Scene Tree

WeirdRubberDuck opened the pull request:
Implements the scene tree part of the scene menu, including a new design that splits the scene graph node view from the tree. 

**OBS! Note that no work has been done on the components for the Properties in this PR, to limit the scope of the changes. I will create another PR with the updated properties** 

To test: 

1. The behavior of the tree, related features, and the new design. 
2. Keyboard navigation!  This needs some improvement, but I would like to discuss how to best implement it
3. Code structure and code itself. Does ti make sense how files are located and split up, etc.. 
4. New features:  Filter by tags, and new information panel for scene graph nodes (including tags and copy-buttons)

Some TODOs (that need discussion and/or could be fixed later):
- [ ] Decide how SGNs should look in other lists (such as the Exoplanet panel and ScreenSpace panel). Might for example need a "Show me this node in the Scene Tree" feature, at least for exoplanets. 
- [x] Fade checkbox implementation is temporary. We should decide on an alternative design that feels more "snappy". 
- [ ] When opening a new window, it should automatically be given focus (to make it easier to navigate using a keyboard). It should also be easy to close (e.g. using ESC ?), and when doing so, return focus to the previous node. 
- [x] Keyboard navigation on the tree - how should it work? E.g. when on a Scene Graph Node, do we want to toggle it or open the window with settings when for example pressing the ENTER key?
- [ ] Virtualize the tree, based on component from https://github.com/OpenSpace/OpenSpace-WebGui/pull/22. It is really slow when all nodes are expanded... (partly because the tree is rendered as soon as an item is expanded). I have started looking at other alternatives a bit, an this seems like a promising option: https://github.com/brimdata/react-arborist
- [ ] Add a way to reorder the layers that's accessible (e.g. using keyboard input)

There are also a lot of style-related things to decide, such as aligning things nicely in the tree and deciding on sizes that make it easier to see the content of the tree. For now, I have tried to keep it as simple as possible so that we can update those things later on.

WeirdRubberDuck commented:
Found some things that needed fixing, so drafting it temporarily

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayersPropertyOwner.tsx:
```diff
@@ -0,0 +1,56 @@
+export function GlobeLayersPropertyOwner({ uri }: Props) {
+  const globeIdentifier = sgnIdentifierFromSubownerUri(uri);
+
+  // TODO: Move icons to the icons file, once decided on the final set and if we use them
```
Do this

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayer.tsx:
```diff
@@ -0,0 +1,53 @@
+          <Tooltip text={propertyOwner?.description || 'No information'} />
+          {showDragHandle && dragHandleProps && (
+            <ThemeIcon variant={'default'} {...dragHandleProps}>
+              <MdDragHandle />
```
Move icon to icons file

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,161 @@
+              withArrow
+            >
+              <ActionIcon variant={'subtle'} onClick={tree.collapseAllNodes}>
+                <FiChevronsUp />
```
Move icon to icons file. Also on next ActionIcon

WeirdRubberDuck commented on src/panels/Scene/SceneTree/sortingUtil.ts:
```diff
@@ -0,0 +1,169 @@
+  return result;
+}
+
+// Sort a list of items in the scene menu tree. This is a bit complicated, since there are
```
What comment style to use for functions? Should this be a doxygen-style comment?

engbergandreas commented:
Suggestion to add `Enter` key to open a SGN and `Space` to toggle it on/off when navigating the scene tree. Also, missing keyboard inputs when focused on on/off checkbox in the tree

engbergandreas commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+  return properties[`${uri}.GuiOrderingNumber`]?.value as number | undefined;
+}
+
+export function isRenderable(uri: string) {
```
```suggestion
export function isRenderable(uri: string): boolean {
```

engbergandreas commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+  );
+}
+
+export function isGlobeLayersUri(uri: string, properties: Properties) {
```
```suggestion
export function isGlobeLayersUri(uri: string, properties: Properties): boolean {
```

engbergandreas commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+
+// Visible means that the object is enabled, based on the values of its enabled and fade
+// properties (which both may be undefined)
+export function checkVisiblity(enabled: boolean | undefined, fade: number | undefined) {
```
```suggestion
export function checkVisiblity(enabled: boolean | undefined, fade: number | undefined): boolean | undefined {
```

engbergandreas commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+// properties (which both may be undefined)
+export function checkVisiblity(enabled: boolean | undefined, fade: number | undefined) {
+  // Enabled is required, but fade can be optional
+  if (enabled === undefined) {
+    return undefined;
```
do we want to return false instead?

WeirdRubberDuck replied:
No, beacuse the result of this check is currently used to determine if the checkbox should be rendered or not.

engbergandreas commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+}
+
+// Returns whether a property matches the current visiblity settings
+export function isPropertyVisible(
+  propertyDetails: PropertyDetails | undefined,
+  visiblitySetting: number | undefined
+) {
```
```suggestion
export function isPropertyVisible(
  propertyDetails: PropertyDetails | undefined,
  visiblitySetting: number | undefined
): boolean {
```

engbergandreas commented on src/windowmanagement/WindowLayout/WindowLayoutContext.tsx:
```diff
@@ -1,11 +1,13 @@
   ref: Ref<DockLayout>;
   addWindow: (component: JSX.Element, options: WindowLayoutOptions) => void;
+  closeWindow: (id: string) => void;
+  createWindowTabData: (id: string, title: string, content: JSX.Element) => TabData;
```
```suggestion
  createWindowTabData: (id: string, title: string, content: React.JSX.Element) => TabData;
```

WeirdRubberDuck commented on src/components/Property/Property.tsx:
```diff
@@ -68,9 +69,7 @@ interface Props {
-export function Property({ uri }: Props) {
-  // TODO: The state here has been changed. Should just be propertyTree.properties. Or
-  // maybe even just state.properties?
+export const Property = React.memo(({ uri }: Props) => {
```
How did you have to write it in your case? 

In this case, there doesn't seem to be any option. Memo requires this syntax

WeirdRubberDuck replied:
However, it should be
```suggestion
export const Property = memo(({ uri }: Props) => {
```

engbergandreas replied:
This is what it looks like in the TimeIndicator: 
``` TypeScript
export const TimeIndicator = forwardRef<SVGRectElement, TimeIndicatorProps>(
  function TimeIndicator(
    { yScale, margin, timelineWidth, scale }: TimeIndicatorProps,
    ref
  ) {
```

engbergandreas replied:
But now that I go back to it, it should probably be something like this instead: 

```JavaScript
 export const TimeIndicator = forwardRef<SVGRectElement, TimeIndicatorProps>(
  (
    { yScale, margin, timelineWidth, scale }: TimeIndicatorProps,
    ref
  )  => {
  ```

WeirdRubberDuck commented on src/components/PropertyOwner/PropertyOwner.tsx:
```diff
@@ -1,65 +1,95 @@
+  withHeader = true,
+  expandedOnDefault = false
+}: Props) {
   const propertyOwner = useAppSelector(
```
Cool, wasn't aware we had one

WeirdRubberDuck commented on src/components/CopyToClipboardButton/CopyToClipboardButton.tsx:
```diff
@@ -0,0 +1,26 @@
+            variant={'subtle'}
+            onClick={copy}
+          >
+            {copied ? <CopyIcon /> : <CopyIcon />}
```
Looks like a woopsie ;)

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayersGroup.tsx:
```diff
@@ -0,0 +1,47 @@
+      <Paper p={'xs'}>
+        <LayerList layers={layers} layerGroup={propertyOwner.identifier} globe={globe} />
+        <Space h={'xs'} />
+        {properties.map((p) => (
+          <Property key={p} uri={p} />
```
This component already has a `uri` variable. We should avoid reusing the name in this case. 

But if it is unclear, it could be renamed to e.g. `pUri` or `propertyUri`. Personally I preferred having it as short as possible here, though, since the variable is so local

engbergandreas replied:
It's not that big of a deal but since `properties` here is a list of strings and not the type `Property` I think that's where my confusion comes from. At first glance, it looks like we're mapping over property objects but we only store the URI:s. But yeah I guess we should decide on that because I know other places where we use the same prop name and argument

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTreeNode.tsx:
```diff
@@ -0,0 +1,42 @@
+import { Box, RenderTreeNodePayload, TreeNodeData } from '@mantine/core';
+
+import { CollapsableHeader } from '../../../components/CollapsableHeader/CollapsableHeader';
```
Haha, lesson from this PR is that I'm relying too much on the automatic creation of imports ;)

WeirdRubberDuck commented on src/panels/Scene/hooks.ts:
```diff
@@ -0,0 +1,60 @@
+  const { ref, addWindow, closeWindow, createWindowTabData } = useWindowLayoutProvider();
+
+  function openCurrentNodeWindow(content: JSX.Element) {
+    if (!ref || !('current' in ref) || ref.current === null) {
```
I was also a bit annoyed by having to add that part, but this is the error I get if I don't: 
![image](https://github.com/user-attachments/assets/494cedb6-dc87-43c0-9f73-3091216194c6)

Haven't found a way to get around it, but am very much open to suggestions

engbergandreas replied:
Problem was that the context provider had the wrong type `Ref` instead of `RefObject` pushing a change for this

WeirdRubberDuck commented on src/redux/documentation/documentationMiddleware.ts:
```diff
@@ -0,0 +1,28 @@
+export const fetchDocumentation = createAsyncThunk(
+  'documentation/fetchDocumentation',
+  async (_, thunkAPI) => {
+    // @TODO (emmbr, 2024-12-04): We could also get the rest of the documentation data
```
I just felt like it was nice to note down what type of information exists in the documentation topic and hence can be accessed through `api.getDocumentation`. Had no specific use case in mind

WeirdRubberDuck replied:
But maybe this is more confusing than helpful, and the comment can be removed

WeirdRubberDuck commented on src/redux/groups/groupsSlice.ts:
```diff
@@ -1,38 +1,62 @@
 const initialState: GroupsState = {
   customGroupOrdering: {},
-  groups: {}
+  groups: {}, // @TODO (emmbr, 2024-12-04): this could maybe be removed from the state
```
If it's used, let's keep it :)  I can imagine that there are more scenarios when we would like to get the information for a certain group

WeirdRubberDuck commented on src/redux/groups/groupsSlice.ts:
```diff
@@ -1,38 +1,62 @@
     setGroups: (state, action: PayloadAction<Groups>) => {
       state.groups = action.payload;
+    },
+    setSceneTreeData: (state, action: PayloadAction<TreeNodeData[]>) => {
+      state.sceneTreeData = action.payload;
+    },
+    setTags: (state, action: PayloadAction<string[]>) => {
+      state.tags = action.payload;
+    },
+    storeSceneTreeNodeExpanded: (state, action: PayloadAction<string[]>) => {
+      state.expandedGroups = action.payload;
+      return state;
```
Good catch on the inconsistencies here. I'll add the `return state;` to the other functions, to be more consistent with the other parts of the code. 

But yeah, we should make a decision, since now we mix between the `return state;` and `return { ... }` version

WeirdRubberDuck commented on src/redux/groups/groupsSliceMiddleware.ts:
```diff
@@ -37,6 +36,8 @@ const computeGroups = (
     groups[guiPath] = groups[guiPath] || emptyGroup();
     groups[guiPath].propertyOwners.push(uri);
+
+    // Finally, sort the property owners based on the GuiOrderingNumber
```
Woopsie :) I guess that I at some point thought that it would be more useful to sort the nodes directly when generating the data for the groups.

It might still be, but if so, when generating the tree data. That's something to discuss I guess. Currently the sorted and filtered tree data is memoized

WeirdRubberDuck commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,125 @@
+
+export function isPropertyOwnerHidden(properties: Properties, uri: string) {
+  const prop = properties[`${uri}.GuiHidden`];
+  if (prop && prop.value) {
+    return true;
+  }
+  return false;
```
The property value `prop.value` should always be a boolean, so this is not a problem here, but I can see how it feels unclear. 

This is a direct rewrite of the old code. A potential alternative way to write it could be: 
```
const isHidden = properties[`${uri}.GuiHidden`]?.value as boolean | undefined;
return isHidden || false;
```

engbergandreas replied:
Ah okay makes sense, I think the alternative is clearer. But I'm not a big fan of us having to explicitly type `as` everywhere since that can lead to bugs down the road. I think at some point we should revisit the `property.value` and see if we can come up with a better solution that does not require us to type cast it whenever we need to use it.

WeirdRubberDuck replied:
I agree with that - i.e. to avoid using "as". 

However, I also feel like this typing business maybe not have to be perfect, as long as it works and is understnadable. To some degree, we also need to balance effort with benefit.

WeirdRubberDuck replied:
But I'll change it according to the suggested alternative

WeirdRubberDuck commented on src/windowmanagement/WindowLayout/WindowLayoutContext.tsx:
```diff
@@ -1,11 +1,13 @@
 
 export interface ProviderProps {
   ref: Ref<DockLayout>;
   addWindow: (component: JSX.Element, options: WindowLayoutOptions) => void;
```
Why? Care to motivate?

WeirdRubberDuck replied:
So far we've only used `JSX.Element` everywhere

engbergandreas replied:
`JSX.Element` is a deprecated hovering it suggests using the `React.JSX.Element` instead so I was just following that.

WeirdRubberDuck replied:
Huh, interesten that it does not warn. Then we should change this _everywhere_.

Or should we really be using `React.ReactElement`?  
https://stackoverflow.com/questions/58123398/when-to-use-jsx-element-vs-reactnode-vs-reactelement

So confused about all this :)

WeirdRubberDuck commented on src/components/PropertyOwner/PropertyOwner.tsx:
```diff
@@ -1,65 +1,95 @@
 
-  const subPropertyOwners = propertyOwner?.subowners || [];
-  const name = propertyOwner?.name;
+  if (!propertyOwner) {
+    throw Error(`No property owner found for uri: ${uri}`);
+  }
```
The throw signals that we shouldn't even get here, so the hook order should not be a problem. 

However, I have forgotten to check and handle the case when a property owner is removed, and that currently leads to crash if this component is rendered due to this throw. Need to find a way to deal with that

WeirdRubberDuck replied:
Note to self: Returning an empty fragment here would mean that the hook order changes, so that is not an option.

WeirdRubberDuck replied:
Ok, so the way I did it now is that I removed the throw if the component does not directly represent a property owner. My thought is that this component should not be added in the first place if the property owner for it does not exists, and that in this case it makes sense to throw

The same goes for globe layers and globe layer groups. In these cases, I kept the throw, in the other components I changed the handling to return an empty fragment or show some meaningful information. It no longer crashes when removing an existing property owner. 

Thoughts?

WeirdRubberDuck commented on src/components/NodeNavigationButton/NodeNavigationButton.tsx:
The reason I changed the sizing for the buttons is that I added the option to create these buttons in different sizes, since they were way too big to use in the Scene Menu tree. And the `IconSize.sm` size does not work well when the `ActionIcon` is smaller, it looks cropped and ugly: 
![image](https://github.com/user-attachments/assets/68d67386-910f-4243-be80-b632f0a55f35)

My thought was that removing the added size was the way to go here, as sizing is part of styling and could be left to the design phase of the project. But I'm open for discussion on how to solve it

engbergandreas replied:
Ah I see, what if we have two sizing props to adjust the Button and the Icon size respectively?

WeirdRubberDuck commented on src/panels/Scene/hooks.ts:
```diff
@@ -0,0 +1,60 @@
+      }
+
+      if (scenePanelParentBox.parent && scenePanelParentBox.parent.mode === 'float') {
+        // @TODO (emmbr, 2024-12-04): If floating window, split the scene tree in two
+        // instead?
```
I don't think we want to do this, due to added complexity. So remove this comment. 

It now pops up as another floating window, which is good enough

WeirdRubberDuck commented on src/panels/Scene/SceneTree/treeUtil.ts:
```diff
@@ -0,0 +1,70 @@
+): TreeNodeData[] {
+  return nodes
+    .map((node) => {
+      const newNode = { ...node };
```
Yes, we do. `node` is read-only, which means that we can not assign values to `node.children`. Hence the copy

WeirdRubberDuck commented on src/panels/OriginPanel/FocusEntry.tsx:
```diff
@@ -47,12 +47,14 @@ export function FocusEntry({
             <NodeNavigationButton
               type={NavigationType.frame}
               identifier={entry.identifier}
+              size={'lg'}
```
See answer to that comment. But it would also be good to discuss it IRL

engbergandreas replied:
I think we can leave it for now and discuss a solution when we start dealing with all other css

WeirdRubberDuck commented on src/panels/Scene/SceneTree/sortingUtil.ts:
```diff
@@ -0,0 +1,169 @@
+    treeSortingInfo: TreeSortingInfo
+  ) {
+    return nodes.map((node) => {
+      const n = { ...node };
```
No, since `node` is read-only and we cannot modify it's children

engbergandreas commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,161 @@
+import { storeSceneTreeNodeExpanded } from '@/redux/groups/groupsSlice';
+import { useAppDispatch, useAppSelector } from '@/redux/hooks';
+
+import { useOpenCurrentSceneNodeWindow } from '../hooks';
```
Ah, no you're right. I thought this was the general hooks import.

WeirdRubberDuck commented on src/panels/Scene/SceneTree/sortingUtil.ts:
```diff
@@ -0,0 +1,169 @@
+  guiOrder: number | undefined;
+}
+
+interface TreeSortingInfo {
+  [key: string]: TreeSortingInfoEntry;
+}
```
Not sure where you think `undefined` should be added here.  And if I understand it correctly, the interface here just defines that the `TreeSortingInfo` is an object that has string key values and values of the type `TreeSortingInfoEntry`, if any are added. 

If the key does not exist, the value will always be undefined, naturally, but I don't see how that relates to the interface here.

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/LayersList.tsx:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  return (
+    <DragDropContext onDragEnd={onDragEnd}>
+      <Droppable droppableId={'layers'}>
+        {(provided) => (
+          <div {...provided.droppableProps} ref={provided.innerRef}>
+            {renderedLayersList.map((layerUri, index) => (
+              <Draggable key={layerUri} draggableId={layerUri} index={index}>
+                {(item) => (
+                  <div ref={item.innerRef} {...item.draggableProps}>
+                    <GlobeLayer
+                      uri={layerUri}
+                      showDragHandle={renderedLayersList.length > 1}
+                      dragHandleProps={item.dragHandleProps}
+                    />
+                  </div>
+                )}
+              </Draggable>
+            ))}
+            {provided.placeholder}
+          </div>
+        )}
+      </Droppable>
```
I would honestly prefer not to, and refer to the documentation of the `hello-pangea/dnd` library. The divs and props are very specific to that library, and I don't really see the point of adding comments that needs maintenance and clutters the code in this case.  

If we want a comment, maybe something like "Create a drag-n-drop reorderable list using the components from hello-pangea/dnd" would suffice? Or renaming the component to make it clearer that it's a reorderable list. 

One thing I don't quite like with the current implementation is that the `GlobeLayer` takes the props for the "drag handle", but when I started implementing it this was a necessity.  Might give this a second try now that the handle is its own icon, and not the entire object, however , and see  if I can keep all the dragging logic in the same component

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/LayersList.tsx:
```diff
@@ -0,0 +1,66 @@
+import { GlobeLayer } from './GlobeLayer';
+
+interface Props {
+  layers: string[]; // List of URIs
```
What were the opinions you guys had about it?   I've thought about the same thing, and it might be suitable to do it as part of this PR since it relies so heavily on uris.

WeirdRubberDuck replied:
Conclusion: We decided to do this. I made a PR for adding the type aliases to master: https://github.com/OpenSpace/OpenSpace-WebGui/pull/26

Should apply it to this PR once that is merged

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeHeader.tsx:
```diff
@@ -0,0 +1,120 @@
+  const { addWindow } = useWindowLayoutProvider();
+  function openInNewWindow() {
+    const content = <SceneGraphNodeView uri={uri} />;
+    addWindow(content, {
+      id: 'sgn-' + uri,
+      title: name,
+      position: 'float'
```
I feel like for that we would need some type of variable for a standard size that the windows get, which depends on the design for different size displays. So I would prefer to leave it for now.

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeMetaInfo.tsx:
```diff
@@ -0,0 +1,137 @@
+          ))}
+        </Group>
+      ],
+      ['GUI:', <span style={{ overflowWrap: 'anywhere' }}>{guiPath}</span>]
```
The reason I didn't is that `Text` adds a different style and size compared to pure text (it's a little bit larger than the default one), so I would have to add it to all the table data items. But it felt weird to make _everything_ a `Text` component, when it's not used everywhere else in the UI. 

But that's something to discuss. If we want to use the `Text` component from Mantine, we should apply it everywhere in the UI

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeMetaInfo.tsx:
```diff
@@ -0,0 +1,137 @@
+      ['Name:', documentation?.name],
+      [
+        'Path:',
+        <span style={{ overflowWrap: 'anywhere' }}>{documentation?.path}</span>,
```
Se comment above

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeView.tsx:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  const propertyOwners = useAppSelector((state) => state.propertyOwners.propertyOwners);
+  const subOwnerDataMap: { [id: string]: PropertyOwnerType } = {};
```
I see your point. Will see if I can rewrite in a more typed way

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeView.tsx:
```diff
@@ -0,0 +1,107 @@
+
+  // We know that all scene graph nodes have the same subowners. However, not all of them
+  // are guaranteed to exist
+  const subowners: {
```
@engbergandreas I rewrote large parts of this component to try to make it less confusing. Please give it a look

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeView.tsx:
```diff
@@ -0,0 +1,107 @@
+  propertyOwner.subowners.forEach((subowner) => {
+    const owner = propertyOwners[subowner];
+    if (owner) {
+      subowners[owner.identifier as keyof typeof subowners] = owner;
```
Not a big fan of this, but unsure how to do it otherwise

engbergandreas replied:
Yeah this is gnarly to read but we can define a type eg., `type TransformType = keyof typeof subowners` and do `subowners[owner.identifier as TransformType] = owner`

WeirdRubberDuck commented on src/redux/groups/groupsSlice.ts:
```diff
@@ -1,38 +1,62 @@
+};
+
+export type CustomGroupOrdering = {
+  [key: string]: string[]; // group paths
```
Think I commented on a similar comment somewhere else, but I would like to discuss this!  Have a hard time understanding why it would be necessary. 

The object should not contain any keys for which the value is undefined, and non-existing keys should already be handled by the object type? Please explain to me how it works if I have misunderstood :)

ylvaselling commented on src/panels/Scene/SceneTree/treeUtil.ts:
```diff
@@ -0,0 +1,71 @@
+): TreeNodeData[] {
+  const filteredNodes = nodes
+    .map((node) => {
+      const newNode = { ...node };
+      if (isGroupNode(newNode)) {
+        // Groups => filter children
+        newNode.children = filterTreeData(
+          newNode.children || [],
+          filter,
+          properties,
+          propertyOwners
+        );
+        if (newNode.children.length === 0) {
+          // Don't show empty groups
+          return null;
+        }
+        return newNode;
+      } else {
+        function shouldShowSceneGraphNode(uri: Uri) {
+          let shouldShow = true;
+          if (filter.showOnlyVisible) {
+            shouldShow &&= isSceneGraphNodeVisible(uri, properties);
+          }
+          if (!filter.showHiddenNodes) {
+            shouldShow &&= !isPropertyOwnerHidden(uri, properties);
+          }
+          if (shouldShow && filter.tags.length > 0) {
+            shouldShow &&= filter.tags.some((tag) =>
+              propertyOwners[newNode.value]?.tags.includes(tag)
+            );
+          }
+          return shouldShow;
+        }
+
+        // PropertyOwners (scene graph nodes), may be filtered out based on settings
+        return shouldShowSceneGraphNode(newNode.value) ? newNode : null;
+      }
+    })
```
Consider breaking this out to its own function. Lots of stuff happening with map and filter which makes it a bit hard to read

ylvaselling commented on src/panels/Scene/SceneTree/treeUtil.ts:
```diff
@@ -0,0 +1,71 @@
+        }
+        return newNode;
+      } else {
+        function shouldShowSceneGraphNode(uri: Uri) {
+          let shouldShow = true;
+          if (filter.showOnlyVisible) {
+            shouldShow &&= isSceneGraphNodeVisible(uri, properties);
+          }
+          if (!filter.showHiddenNodes) {
+            shouldShow &&= !isPropertyOwnerHidden(uri, properties);
+          }
+          if (shouldShow && filter.tags.length > 0) {
+            shouldShow &&= filter.tags.some((tag) =>
+              propertyOwners[newNode.value]?.tags.includes(tag)
+            );
+          }
+          return shouldShow;
+        }
```
Consider breaking this out to its own function not within this scope

ylvaselling commented on src/redux/documentation/documentationSlice.ts:
```diff
@@ -0,0 +1,38 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+export interface AssetMetaData {
+  author: string;
+  description: string;
+  identifiers?: string[];
+  license: string;
+  name: string;
+  path: string;
+  url: string;
+  version: string;
+}
```
Place this in types file?

ylvaselling commented on src/redux/groups/groupsSliceMiddleware.ts:
```diff
@@ -61,6 +60,71 @@ const computeGroups = (
   return groups;
 };
 
+export function treeDataForPropertyOwner(uri: Uri, propertyOwners: PropertyOwners) {
+  const propertyOwner = propertyOwners[uri];
+  return {
+    label: propertyOwner?.name || '',
+    value: uri
+  };
+}
```
Since this function is used in other places as well I would think it belongs in some kind of utils file. We might want to consider moving all these to a utils file actually. It is a bit hard to see what this middleware does since there are so many helper functions

WeirdRubberDuck replied:
👍 Moving this and `treeDataFromGroups` to treeUtil.ts. That way, at least all functions which are related to creating the `ThreeNodeData` data structure used for rendering the Mantine tree are removed from this file.  

The reason I put them here to begin with is that I preferred to have all code for generating the data in the same file, to make it easier to overview or find out how the resulting data looks like. For that reason, it also makes sense for me to _not_ move the `computeGroups` and `collectExistingTags` functions somewhere else. But as usual, open for discussion

ylvaselling commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,127 @@
+}
+
+export function isRenderable(uri: Uri): boolean {
+  const renderableSuffix = '.Renderable';
```
All these strings could be keys in the keys file

ylvaselling commented on src/util/propertyTreeHelpers.ts:
```diff
@@ -1,7 +1,127 @@
+    return true;
+  }
+
+  function propertyVisibilityNumber(visibility: PropertyVisibility) {
+    switch (visibility) {
+      case 'Hidden':
+        return 5;
+      case 'Developer':
+        return 4;
+      case 'AdvancedUser':
+        return 3;
+      case 'User':
+        return 2;
+      case 'NoviceUser':
+        return 1;
+      case 'Always':
+        return 0;
+      default:
+        return 0;
+    }
+  }
```
Consider making this an enum

ylvaselling commented on src/windowmanagement/WindowLayout/WindowLayoutProvider.tsx:
```diff
@@ -75,11 +68,39 @@ export function WindowLayoutProvider({ children }: { children: React.ReactNode }
+    }
+  }
+
+  function createWindowTabData(
+    id: string,
+    title: string,
+    content: React.JSX.Element
+  ): TabData {
+    return {
+      id,
+      title,
+      content: <ScrollArea h={'100%'}>{content}</ScrollArea>,
+      closable: true,
+      cached: true,
+      group: 'regularWindow'
+    };
+  }
+
```
This seems like a pure function to me. Could we make this a helper function and not have it as a part of this provider?

WeirdRubberDuck replied:
We could.  The reason I put it here for now is that the use case for me felt similar to the `addWindow` function, for example, and that it directly controls the code of the windows that the provider creates. So to me, it made sense to have it in the same class. 

Would be good to discuss where to place it, more precisely :)

WeirdRubberDuck commented on src/components/Collapse/CollapsableContent/CollapsableContent.tsx:
```diff
@@ -0,0 +1,43 @@
+  // The title of the collapsable content.
+  title: React.ReactNode;
+  // If true, the content will not transition in or out. Helps with performance.
+  noTransition?: boolean;
```
Usually I would agree with this, but here there is another consideration which is that the prop can be specified like this 
```
<CollapsableContent noTransition />
```
i.e. without specifying the actual boolean value when it is set to true, and not set it at all if it is false. 

So, here the negative phrasing is intentional, and the double negation situation would never really occur

WeirdRubberDuck commented on src/components/PropertyOwner/VisiblityCheckbox.tsx:
```diff
@@ -0,0 +1,65 @@
+export function PropertyOwnerVisibilityCheckbox({ uri, onChange }: Props) {
+  const luaApi = useOpenSpaceApi();
+
+  const enabledUri = `${uri}.Enabled`;
+  const fadeUri = `${uri}.Fade`;
```
I don't think we should add the specific property identifiers as keys in the file, that seems a bit "overkill" to me.  

However, we could potentially add helper functions to create the string for the property URI from the URI of a renderable. For example: 
```typescript
function fadeUri(uri: Uri) {
  return `${uri}.Fade`;
}
```
and maybe add that to the propertyTreeHelper.ts file. 

but I don't feel 100% convinced that that would make the code easier to read/understand

WeirdRubberDuck replied:
I went with this approach, but let's have a deeper discussion on this and how we feel about where to use keys once @engbergandreas is back

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayersGroup.tsx:
```diff
@@ -0,0 +1,46 @@
+    >
+      <Paper p={'xs'}>
+        <LayerList layers={layers} layerGroup={propertyOwner.identifier} globe={globe} />
+        <Space h={'xs'} />
```
In this case, the component above is not a Mantine component, so we can't add the margin directly to that one. So,  I think this counts as the case for which this component was intended, according to this info: https://mantine.dev/core/space/#where-to-use

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayersPropertyOwner.tsx:
```diff
@@ -0,0 +1,55 @@
+export function GlobeLayersPropertyOwner({ uri }: Props) {
+  const globeIdentifier = sgnIdentifierFromSubownerUri(uri);
+
+  const groups = [
+    {
+      id: 'ColorLayers',
+      icon: <ColorPaletteIcon />
+    },
+    {
+      id: 'HeightLayers',
+      icon: <LandscapeIcon />
+    },
+    {
+      id: 'NightLayers',
+      icon: <NightIcon />
+    },
+    {
+      id: 'Overlays',
+      icon: <LayersIcon />
+    },
+    {
+      id: 'WaterMasks',
+      icon: <WaterIcon />
+    }
+  ];
```
Personally, I prefer keeping things that are only used in one place as locally as possible in the code and I think this counts as one of those cases. Moving it somewhere else in this case would make the code more difficult to read, IMO. 

But if we have a good place where we define which icons are used where, I could maybe be convinced otherwise. Let me know if you have a suggestions as for where to move it

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeMetaInfo.tsx:
```diff
@@ -0,0 +1,136 @@
+}
+
+export function SceneGraphNodeMetaInfo({ uri }: Props) {
+  const identifier = identifierFromUri(uri);
+
+  let [description] = useGetStringPropertyValue(`${uri}.GuiDescription`);
+  if (description) {
+    description.replace(/\\n/g, '').replace(/<br>/g, '');
+  } else {
+    description = 'No description found';
+  }
+
+  const [guiPath] = useGetStringPropertyValue(`${uri}.GuiPath`);
+
+  const documentation = useAppSelector((state) => {
+    if (identifier) {
+      return state.documentation.assetsMetaData.find(
+        (doc) => doc.identifiers && doc.identifiers.includes(identifier)
+      );
+    }
+    return undefined;
+  });
+
+  const propertyOwner = useGetPropertyOwner(uri);
```
Would be great to have this defined somewhere :)  I wonder about the order every time

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeMetaInfo.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  const propertyOwner = useGetPropertyOwner(uri);
+
+  const mainTableData: TableData = {
```
Similarly as for `GlobeLayersPropertyOwner.tsx`, I think I would prefer not to.  Moving this to another file would make it more difficult to understand what information is being rendered for this component these parts are split up into different files. 

In this case, the variables also depend on the input properties and  would have to be a function or hook or something if moved. 

But, open for discussion, of course

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeView.tsx:
```diff
@@ -0,0 +1,115 @@
+    );
+  }
+
+  const TabKeys = {
+    Renderable: 'Renderable',
+    Transform: 'Transform',
+    Other: 'Other',
+    Info: 'Info'
+  };
```
Enum, yes (forgot that stringed enums was a thing, as it isn't in C++ 🙂)

However, I would not want to move it to the enums file as it is only relevant for this specific component and so directly tied to it.  As usual, open for discussion though

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,160 @@
+          // Only add leaf nodes to the flat data
+          flatData.push({
+            ...node,
+            guiPath: properties[`${node.value}.GuiPath`]?.value as string | undefined
```
As mentioned previously, I would prefer to not use keys for specific property names as it makes the code more difficult to read.  But include this as part of the discussion when @engbergandreas is back

However, again we could make a helper function to get the value here, to make the code more readable.

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,160 @@
+    return data;
+  }, [customGuiOrdering, filter, properties, propertyOwners, sceneTreeData]);
+
+  function flattenTreeData(treeData: TreeNodeData[]): SearchData[] {
```
Could move it to treeHelper.ts 👍

WeirdRubberDuck commented on src/panels/Scene/SceneTree/treeUtil.ts:
```diff
@@ -0,0 +1,71 @@
+  return isGroup(node.value);
+}
+
+export interface SceneTreeFilterProps {
+  showOnlyVisible: boolean;
+  showHiddenNodes: boolean;
+  tags: string[];
+}
```
Ah, the naming here is a remnant. It used to be an interface that was directly used as the props for `SceneTree` and `FeaturedSceneTree`, and apparently I didn't think enough when I did the renaming 🙂  How about calling it `SceneTreeFilterSettings` ?

I would not want to move it to the big general types file, as it is a type that is only relevant for the SceneTree feature.  However, discussed a bit with @engbergandreas about maybe starting to have more smaller local types files and this could be one case where that would be relevant, maybe

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,161 @@
+      // Also close the "current node" window
+      closeCurrentNodeWindow();
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```
We're not wuite sure what you mean with memoing the function here :)

WeirdRubberDuck replied:
Decided to include the funciton in the dependency list instead.

WeirdRubberDuck replied:
Nvm it didn't work :( The component rerenders too often then

engbergandreas replied:
I'm not sure what past me meant either but I guess it has to do with adding it to the dependency array. Which function causes the rerenders? Could we memo it?

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,160 @@
+            tree={tree}
+            renderNode={(payload) => <SceneTreeNodeStyled {...payload} />}
+          />
+          <Group gap={0} pos={'absolute'} top={0} right={0}>
```
I get that comment. Here, this group is accessed after you have stepped through the tree, which I can personally find a bit weird. It would bake more sense if it is accessed before the tree itself. 

Do you have any thoughts on an alternative placement/implementation in this case?  I put the group like this to maximize the amount of space that the tree gets

WeirdRubberDuck commented on src/panels/Scene/SceneTree/SceneTree.tsx:
```diff
@@ -0,0 +1,160 @@
+
+  // We use a ref here, since we need this object to exists for the entire lifetime of the
+  // component, including when the component is unmounted
+  const expandedGroups = useRef<string[]>(initialExpandedNodes);
```
No, but they are still stored _during_  the unmount, which is what matters here. 

In this case, the ref is not used for rendering, only to keep track of which nodes are currently expanded. So hopefully, it's not an issue in this case

WeirdRubberDuck commented on src/panels/Scene/SceneGraphNode/SceneGraphNodeHeader.tsx:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  return (
+    <ThreePartHeader
```
I rewrote the code slightly so that we avoid passing large components a sprops, but couldnät come up with a nice alternative way to implement the `ThreePartHeader` and `Collapsable`. 

Leaving them as is for now, since the functionality is dependent on the three-part slit of the header to handle button clicks on the central part separately from the left and right parts

WeirdRubberDuck commented on src/panels/Scene/GlobeLayers/GlobeLayer.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  return (
+    <CollapsableContent
+      title={<Text c={textColor}>{displayName(propertyOwner)}</Text>}
+      leftSection={
+        <PropertyOwnerVisibilityCheckbox
+          uri={uri}
+          onChange={(isChecked) => setIsActive(isChecked)}
+        />
+      }
+      rightSection={
+        <Group wrap={'nowrap'}>
+          <Tooltip text={propertyOwner.description || 'No information'} />
+        </Group>
+      }
```
See other comment : https://github.com/OpenSpace/OpenSpace-WebGui/pull/23#discussion_r1906828176

WeirdRubberDuck commented:
> Overall it looks awesome!! The popping out of the scenegraph node and the different tabs are 💯
> 
> Some thoughts:
> 
> 1. Not sure this is within the scope of this pr but a note on the layers:
>    ![image](https://private-user-images.githubusercontent.com/24265984/400806074-7763c134-ee0a-4506-9374-4eb0ce785086.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczNjg3MzksIm5iZiI6MTczNzM2ODQzOSwicGF0aCI6Ii8yNDI2NTk4NC80MDA4MDYwNzQtNzc2M2MxMzQtZWUwYS00NTA2LTkzNzQtNGViMGNlNzg1MDg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDEwMjAzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwZjk1ZTkyMDFiODZjYjg5OWZiOTI4MDI0MzQxMWYxMjA2ZmM3MGVjNjIxZDJiODM5ZDgwODZhMDk3MGRmNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RQYPqSWs8E_wmogNH_BhFh6IWAE_FHBe65CD-B68Ces)
>    To me, the not selected layers look disabled. Maybe don't gray them out?
> 2. When filtering the scene menu with tags, maybe we should expand all as default? Seems that you want to do that every time since you want to see what is still in the list
> 3. The "expand all" and "collapse all" are on the same line as the "Solary System" title which is a bit confusing. Is there any way to separate them so that it is clear that collapse and expand are for everything?
> 4. Should the tag filter apply to the quick access? I think maybe not?
> 5. Since we search in the scene menu, maybe the Current focus and the Quick access should actually be above the input field? maybe a big redesign but perhaps worth a thought
> 6. LOVE the uri copy tag. Make the URI "code" styled?
> 7. There is a thin line under the tabs in the scene graph node menu
> 8. Enter doesn't work to turn on / off a layer, only Space. Neither Enter nor Space works to turn off / on a scene graph node.
> 9. Infoboxes are not focusable so not accessible
> 10. Maybe not within the scope of this PR but layer reordering is not accessible (at least not to my knowledge)
> 11. Clicking enter on the Current Focus leads to an empty window for the Scene: Selected Node
> 12. Maybe Space and Enter should both toggle the tree? Space now toggles if the whole line is selected, whereas Enter toggles only on the arrow. Maybe both Space and Enter should toggle on both the whole line and the arrow?
> 13. The "collapse all" and "expand all" come last in the "tabbing hierarchy". They should probably come first, before the scene tree.
> 14. The focus border in the scene tree is not visible on the left hand side and the right hand side

1. Done!  Made it green/white instead of white/gray
2. We discussed this, and my suggestion is to save improvements for the Scene Tree for another PR. 
3. They are still on the same line, to utilize screen space. However, they now come first in the tab order. 
4. I agree - changed so it is not applied, and moved things around so that it's clearer. 
5. Did this, as part of the restructuring for 4. 
6. Done! 
7. This is part of the default style for that Mantine Tabs variant. So left it as is
8. Fixed
9. Not adressed as part of this PR, since it just uses the existing InfoBox component. We should make them toggleable on touch/mouseclick as part of a separate PR. 
10. Leaving this as a future TODO.  For now I just made it work the same way as in the existing UI
11. This was due to the Renderable tab begin selected and then the focus was changed to a node that does not have a renderable.  Changed it so the tab is always visible, but disabled for nodes without a renderable. This prevents the empty window 
12. We discussed this, and I agree. Have not fixed it yet though, so its a TODO still
13. Fixed. They are now before the tree in the tabbing hierarchy
14. Not fixed - but the reason is that there is not enough margin for the focus indicator to the left. This should maybe be part  of a general styling pass for the focus indicator. Currently, it looks different for Mantine components and other components

ylvaselling commented:
Great!
